### PR TITLE
Rule S1151: "switch case" clauses should not have too many lines

### LINKED ISSUE #81: Rule S1151: "switch case" clauses should not have too many lines of code
Implements [RSPEC-1151](https://jira.sonarsource.com/browse/RSPEC-1151)

### PR DESCRIPTION
**Pull request by stylianos-agapiou-sonarsource**
fixes #81

### CODE REVIEW

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/SwitchCaseTooBigCheck.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    defaultValue = "" + DEFAULT_MAXIMUM)
+  public int max = DEFAULT_MAXIMUM;
+
+  public SwitchCaseTooBigCheck() {
+    super(UastNode.Kind.CASE, UastNode.Kind.DEFAULT_CASE);
+  }
+
+  @Override
+  public void visitNode(UastNode node) {
+    int caseLength = node.lastToken().line - node.firstToken().line;
```
Comment (vilchik-elena) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/SwitchCaseTooBigCheck.java]: @stylianos-agapiou-sonarsource don't count line for `case condition`

File: common-rule-engine/src/test/files/checks/SwitchCaseTooBigCheck/SwitchCaseTooBigCheck.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
+package SwitchCaseTooBigCheck
```
Comment (vilchik-elena) [inline common-rule-engine/src/test/files/checks/SwitchCaseTooBigCheck/SwitchCaseTooBigCheck.go:1]: @stylianos-agapiou-sonarsource simplify tests

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/SwitchCaseTooBigCheck.java [hunk 3]
```diff
@@ -40,11 +41,14 @@ public SwitchCaseTooBigCheck() {
...
-      node.getChild(UastNode.Kind.KEYWORD).ifPresent(caseKeyword -> reportIssue(caseKeyword,
-        "Reduce this switch case number of lines from " + caseLength + " to at most " + max
-          + ", for example by extracting code into methods."));
+    UastNode caseBody = CaseLike.from(node).body();
+    if (caseBody != null) {
+      int caseLength = caseBody.lastToken().line - caseBody.firstToken().line + 1;
+      if (caseLength > max) {
+        node.getChild(UastNode.Kind.KEYWORD).ifPresent(caseKeyword -> reportIssue(caseKeyword,
+          "Reduce this switch case number of lines from " + caseLength + " to at most " + max
+            + ", for example by extracting code into methods."));
```
Comment (vilchik-elena) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/SwitchCaseTooBigCheck.java]: @stylianos-agapiou-sonarsource `methods` -> `functions`

File: common-rule-engine/src/test/files/checks/SwitchCaseTooBigCheck/SwitchCaseTooBigCheck.go [hunk 4]
```diff
@@ -15,48 +14,38 @@ func main(){
...
+			bar()
+			bar()
+		case 6:	// Noncompliant  {{Reduce this switch case number of lines from 7 to at most 5, for example by extracting code into methods.}}
 			if a == 1{
 				fmt.Println(a)
 				fmt.Println(a)
 				fmt.Println(a)
 			}else{
 			fmt.Println(a)
 		}
```
Comment (vilchik-elena) [inline common-rule-engine/src/test/files/checks/SwitchCaseTooBigCheck/SwitchCaseTooBigCheck.go]: @stylianos-agapiou-sonarsource badly formatted

File: common-rule-engine/src/test/files/checks/SwitchCaseTooBigCheck/SwitchCaseTooBigCheck.go [hunk 5]
```diff
@@ -15,48 +14,38 @@ func main(){
...
-		case 1:		// compliant
+		case 1:		// Compliant
 			fmt.Println("OS X1.")
 			fmt.Println("OS X2.")
 			fmt.Println("OS X3.")
 
 
 
-        case 2:	 // Noncompliant {{Reduce this switch case number of lines from 10 to at most 5, for example by extracting code into methods.}}
+        case 2:	 // Noncompliant {{Reduce this switch case number of lines from 7 to at most 5, for example by extracting code into methods.}}
```
Comment (vilchik-elena) [inline common-rule-engine/src/test/files/checks/SwitchCaseTooBigCheck/SwitchCaseTooBigCheck.go]: @stylianos-agapiou-sonarsource and this one badly formatted

Comment (stylianos-agapiou-sonarsource) [reply common-rule-engine/src/test/files/checks/SwitchCaseTooBigCheck/SwitchCaseTooBigCheck.go]: Thanks for the notice, I just formatted the go code with the go fmt tool.

File: sonar-go-plugin/src/main/resources/org/sonar/l10n/go/rules/go/S1151.html [hunk 6]
```diff
@@ -0,0 +1,37 @@
+<p>The <code>switch</code> statement should be used only to clearly define some new branches in the control flow. As soon as a <code>case</code>
+clause contains too many statements this highly decreases the readability of the overall control flow statement. In such case, the content of the
+<code>case</code> clause should be extracted into a dedicated method.</p>
+<h2>Noncompliant Code Example</h2>
+<p>With the default threshold of 3:</p>
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/main/resources/org/sonar/l10n/go/rules/go/S1151.html]: @stylianos-agapiou-sonarsource regenerate

File: common-rule-engine/src/test/java/org/sonar/commonruleengine/checks/SwitchCaseTooBigCheckTest.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+package org.sonar.commonruleengine.checks;
+
+import org.junit.jupiter.api.Test;
+
+import static org.sonar.commonruleengine.checks.TestUtils.checkRuleOnGo;
+
+public class SwitchCaseTooBigCheckTest {
+  @Test
+  void test_go() throws Exception {
+    checkRuleOnGo(new SwitchCaseTooBigCheck());
```
Comment (vilchik-elena) [inline common-rule-engine/src/test/java/org/sonar/commonruleengine/checks/SwitchCaseTooBigCheckTest.java:29]: @stylianos-agapiou-sonarsource add test with not default parameter value

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

**Review by vilchik-elena — APPROVED** [review]

**Pull request merged**